File: 107.二叉树的层次遍历-ii.java
/*
 * @lc app=leetcode.cn id=107 lang=java
 *
 * [107] 二叉树的层次遍历 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

import java.util.LinkedList;
import java.util.Collections;
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if (root == null) 
            return res;
        LinkedList<TreeNode> queue = new LinkedList<>();
        List<Integer> level = new LinkedList<>();
        queue.add(root);
        queue.add(null);
        while (queue.size() > 0) {
            TreeNode node = queue.poll();
            if (node == null) {
                res.add(level);
                level = new LinkedList<>();
                if (queue.size() > 0)
                    queue.add(null);
            } else {
                level.add(node.val);
                if (node.left != null)
                    queue.add(node.left);
                if (node.right != null)
                    queue.add(node.right);
            }
        }
        Collections.reverse(res);
        return res;
    }
}
// @lc code=end

File: 107.二叉树的层次遍历-ii.py
#
# @lc app=leetcode.cn id=107 lang=python3
#
# [107] 二叉树的层次遍历 II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque


class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
        res = []
        que = deque()
        que.append(root)
        que.append(None)
        level = []
        while len(que) > 0:
            node = que.popleft()
            if node is None:
                res.append(level)
                level = []
                if len(que) > 0:
                    que.append(None)
            else:
                level.append(node.val)
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
        return list(reversed(res))

# @lc code=end
